### ISSUE TITLE
`schema_json` can't encode default value of`dict[UUID, Any]` and raises `TypeError`

### ISSUE THREAD

**Issue by Bobronium** [pydantic/pydantic#4774]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When a model has default value of `{uuid4(): "..."}`, and being a dafault value of another model, `json_schema` fails to generate schema with an error.

### Example Code

```Python
from uuid import uuid4

from pydantic import BaseModel
from pydantic import UUID4


class Service(BaseModel):
    urls: dict[UUID4, str] = {uuid4(): "https://example.com"}


class Settings(BaseModel):
    service: Service = Service()


print(Settings.schema_json(indent=4, ensure_ascii=False))
```

```py
Traceback (most recent call last):
  File "/Users/bobronium/Library/Application Support/JetBrains/PyCharm2022.3/scratches/scratch_131.py", line 15, in <module>
    print(Settings.schema_json(indent=4, ensure_ascii=False))
  File "pydantic/main.py", line 674, in pydantic.main.BaseModel.schema_json
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  [Previous line repeated 1 more time]
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 376, in _iterencode_dict
    raise TypeError(f'keys must be str, int, float, bool or None, '
TypeError: keys must be str, int, float, bool or None, not UUID
```

### This works, though

```py
from uuid import uuid4

from pydantic import BaseModel
from pydantic import UUID4


class Service(BaseModel):
    urls: dict[UUID4, str] = {uuid4(): "https://example.com"}


class Settings(BaseModel):
    service: Service  # <--------- Notice no default value here


print(Settings.schema_json(indent=4, ensure_ascii=False))
```
<details>

```json
{
    "title": "Settings",
    "type": "object",
    "properties": {
        "service": {
            "$ref": "#/definitions/Service"
        }
    },
    "required": [
        "service"
    ],
    "definitions": {
        "Service": {
            "title": "Service",
            "type": "object",
            "properties": {
                "urls": {
                    "title": "Urls",
                    "default": {
                        "ac762f77-1e77-4ff8-906a-fd63ca622938": "https://example.com"
                    },
                    "type": "object",
                    "additionalProperties": {
                        "type": "string"
                    }
                }
            }
        }
    }
}
```

</details>

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: .venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.1 (main, Jan 31 2022, 09:16:52) [Clang 13.0.0 (clang-1300.0.29.3)]
                     platform: macOS-12.4-arm64-arm-64bit
     optional deps. installed: ['dotenv', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [X] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Thanks for reporting, if there's an easy fix, happy to accept a PR for 1.10.3.

Otherwise, let's make sure we get this right in v2. Tbh, there are lots of types that currently break JsonSchema.

**Comment by hramezani** [comment]
Fixed in https://github.com/pydantic/pydantic/commit/18359d79853f57a5a892cace2e4e286dd159a300

Cherry-picked into main in https://github.com/pydantic/pydantic/commit/8b7c56879442d5685226d4dab7b5e31e1cd31cc7
